iotcore/src/main/java/com/google/android/things/iotcore/IotCoreClient.java

```diff
@@ -466,25 +537,37 @@ public void connectionLost(Throwable cause) {
 
             @Override
             public void messageArrived(String topic, MqttMessage message) {
-                if (!mConnectionParams.getConfigurationTopic().equals(topic)
-                        || onConfigurationListener == null
-                        || onConfigurationExecutor == null) {
-                    return;
+                if (mConnectionParams.getConfigurationTopic().equals(topic) && onConfigurationListener != null && onConfigurationExecutor != null) {
+                    // Call the client's OnConfigurationListener
+
+                    final byte[] payload = message.getPayload();
+                    onConfigurationExecutor.execute(
+                            new Runnable() {
+                                @Override
+                                public void run() {
+                                    onConfigurationListener.onConfigurationReceived(payload);
+                                }
+                            });
+                } else if (topic.startsWith(mConnectionParams.getCommandsTopic()) && onCommandListener != null && onCommandExecutor != null) {
+                    // Call the client's OnCommandListener
+
+                    final byte[] payload = message.getPayload();
+                    final String prefix = mConnectionParams.getCommandsTopic() + "/";
```

Comment:
"prefix" is supposed to be "^" + mConnectionParams.getCommandsTopic() + "/?".
"^" is not necessary because the given topic is sure to start with mConnectionParams.getCommandsTopic(). but, to express the logic clearly, I recommend to add "^" at the beginning.
commands topic pattern would be:
   - /devices/<deviceID>/commands
   - /devices/<deviceID>/commands/subFolder
so, we are not sure that the topic has always trailing "/". that's why we need to add "?" at the end of regular pattern.

Reply:
nice: good catch, indeed! I will adjust accordingly.